python/lsst/daf/butler/queries/_query.py

```diff
@@ -0,0 +1,652 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Query",)
+
+from collections.abc import Iterable, Mapping, Set
+from types import EllipsisType
+from typing import Any, final, overload
+
+from lsst.utils.iteration import ensure_iterable
+
+from .._dataset_type import DatasetType
+from .._storage_class import StorageClassFactory
+from ..dimensions import DataCoordinate, DataId, DataIdValue, DimensionGroup
+from ..registry import DatasetTypeError, MissingDatasetTypeError
+from ._base import HomogeneousQueryBase
+from ._data_coordinate_query_results import DataCoordinateQueryResults
+from ._dataset_query_results import (
+    ChainedDatasetQueryResults,
+    DatasetQueryResults,
+    SingleTypeDatasetQueryResults,
+)
+from ._dimension_record_query_results import DimensionRecordQueryResults
+from .convert_args import convert_where_args
+from .driver import QueryDriver
+from .expression_factory import ExpressionFactory
+from .result_specs import DataCoordinateResultSpec, DatasetRefResultSpec, DimensionRecordResultSpec
+from .tree import DatasetSearch, InvalidQueryError, Predicate, QueryTree, make_identity_query_tree
+
+
+@final
+class Query(HomogeneousQueryBase):
```

Comment:
Yeah I buy the argument for having `where` on both, and one base class with one method is plenty straightforward.

Reply:
Dropping `DatasetQueryResults` and `ChainedDatasetQueryResults` changes the picture here a bit, and I'm now inclined to punt this proposal to another later ticket:

- We're now down to one base class already.
- When we add a dataset-types-in-collections summary interface ([DM-43157](https://rubinobs.atlassian.net/browse/DM-43157)), we'll have something else we might want to call a "result summary"; it might even just be an expanded version of this proposal that knows how to group by dataset type and/or collection.
- I think we have a _decent_ story right now: you don't need to specify the result type to know whether there will be any results, so `Query` can tell you; for anything beyond that you have to use `QueryResult`.  It's not a great story, because you have to ponder a bit about duplication to understand why `Query` can't at least do `count`.  But I think it isn't terrible.

Reply:
That seems reasonable.